pkg/remediator/watch/filteredwatcher.go

```diff
@@ -258,39 +299,59 @@ func (w *filteredWatcher) Run(ctx context.Context) status.Error {
 		eventCount := 0
 		ignoredEventCount := 0
 		klog.V(2).Infof("(Re)starting watch for %s at resource version %q", w.gvk, resourceVersion)
-		for event := range w.base.ResultChan() {
-			w.pruneErrors()
-			newVersion, ignoreEvent, err := w.handle(ctx, event)
-			eventCount++
-			if ignoreEvent {
-				ignoredEventCount++
-			}
-			if err != nil {
-				if isExpiredError(err) {
-					klog.Infof("Watch for %s at resource version %q closed with: %v", w.gvk, resourceVersion, err)
-					// `w.handle` may fail because we try to watch an old resource version, setting
-					// a watch on an old resource version will always fail.
-					// Reset `resourceVersion` to an empty string here so that we can start a new
-					// watch at the most recent resource version.
-					resourceVersion = ""
-				} else if w.addError(watchEventErrorType + errorID(err)) {
-					klog.Errorf("Watch for %s at resource version %q ended with: %v", w.gvk, resourceVersion, err)
+		eventCh := w.base.ResultChan()
+	EventHandler:
+		for {
+			select {
+			case <-ctx.Done():
+				runErr = status.InternalWrapf(ctx.Err(), "remediator watch stopped for %s", w.gvk)
+				// Stop the watcher & return the status error
+				break Watcher
+			case event, ok := <-eventCh:
+				if !ok { // channel closed
+					// Restart the watcher
+					break EventHandler
+				}
+				w.pruneErrors()
+				newVersion, ignoreEvent, err := w.handle(ctx, event)
+				eventCount++
+				if ignoreEvent {
+					ignoredEventCount++
+				}
+				if err != nil {
+					if errors.Is(err, context.Canceled) || isContextCancelledStatusError(err) {
+						// The error wrappers are especially confusing for
+						// users, so just return context.Canceled.
+						runErr = status.InternalWrapf(context.Canceled, "remediator watch stopped for %s", w.gvk)
+						// Stop the watcher & return the status error
+						break Watcher
```

Comment:
Is there a reason for using these labeled breaks with an extra variable to track the return value, rather than using `return`? It seems like the latter may be more readable and easier to follow.

Reply:
It avoids needing to duplicate the logging.

Reply:
Also even if we use return to escape the function we would still need a named loop to break the event loop, and not the inner select.